### PR TITLE
9 - Family Member Registration

### PR DESCRIPTION
**Pull request by magas4891**

### CODE REVIEW

File: app/controllers/family_connections_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+class FamilyConnectionsController < ApplicationController
+  def index
+    receiver_user = current_user.receiver_user_family_connections
+
+    render_response(receiver_user.map { |connection| connection.present.family_connection_page_context })
+  end
+
+  def update
+    result = Connections::FamilyConnections::UpdateService.perform(update_params)
```
Comment (drkmen) [inline app/controllers/family_connections_controller.rb]: rename `result` to `family_connection`

File: app/controllers/family_connections_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    render_success_response({ connection: result.present.family_connection_page_context }, :accepted)
+  end
+
+  private
+
+  def update_params
+    { connection: current_connection }
+  end
```
Comment (drkmen) [inline app/controllers/family_connections_controller.rb]: permit?

File: app/controllers/users/invitations_controller.rb [hunk 3]
```diff
@@ -5,9 +5,13 @@ class InvitationsController < ApplicationController
     skip_before_action :authenticate_user!, only: :update
 
     def create
-      user = Users::InvitationService.perform(invite_params.merge(current_user: current_user))
-
-      render_success_response({ invite_token: user.raw_invitation_token, user: user.present.page_context }, :created)
+      if receiver_user.present?
+        result = Connections::FamilyConnections::CreateService.perform(family_connection_params)
```
Comment (drkmen) [inline app/controllers/users/invitations_controller.rb]: same here
variable name should contain as much information as possible

File: app/controllers/users/invitations_controller.rb [hunk 4]
```diff
@@ -38,5 +43,16 @@ def accept_invitation_params
...
+
+    def family_connection_params
+      params.require(:user).permit(:member_type)
+            .merge(requestor_user: current_user,
+                   receiver_user: receiver_user)
+    end
+
+    def receiver_user
+      @receiver_user ||= User.where('phone = :phone or email = :email', { phone: params[:user][:phone],
+                                                                          email: params[:user][:email] }).first
```
Comment (drkmen) [inline app/controllers/users/invitations_controller.rb]: create scope for that

File: app/services/connections/family_connections/update_service.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        update_family
+        update_family_connection
+      end
+
+      private
+
+      def update_family
+        requester = User.find(connection['requestor_user_id'])
+        receiver = User.find(connection['receiver_user_id'])
+        member_types = connection['member_type']
```
Comment (drkmen) [inline app/services/connections/family_connections/update_service.rb]: do not hesitate to create separate methods for these

File: app/services/connections/family_connections/update_service.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        receiver.update_column(:member_type, member_types)
+        receiver.family.users.each do |member|
+          requester.family.users << member
+        end
+      end
+
+      def update_family_connection
+        connection.update_column(:accepted, true)
+        connection
```
Comment (drkmen) [inline app/services/connections/family_connections/update_service.rb]: why do you need to return connection here?

File: app/services/family_members/accept_invitation.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      raise ArgumentError, 'Invitation token is missing' if invitation_token.nil?
+
+      User.accept_invitation!(accept_invitation_params)
+    end
+
+    private
+
+    def accept_invitation_params
+      params
+    end
```
Comment (drkmen) [inline app/services/family_members/accept_invitation.rb]: redundant

File: spec/controllers/payments_controller_spec.rb [hunk 8]
```diff
@@ -9,12 +9,14 @@
...
     let(:send_request) { post :create, params: payment_params.merge(user: user), format: :json }
+    let(:send_invalid_request) { post :create, params: payment_params.merge(user: young_user), format: :json }
     let(:payments_params) { ActionController::Parameters.new(payment_params) }
     let(:card_info) { { number: '4000 00100 0000 2230', exp_month: '2', exp_year: '2021', cvc: '314' } }
-    let(:payment_params) { { type: 'card', card: card_info } }
-    let(:user_payment_servise) do
+    let(:payment_params) { {   type: 'card', card: card_info } }
+    let(:user_payment_service) do
       payments_params.permit(:type, card: %i[number exp_month exp_year cvc]).merge(user: user)
     end
```
Comment (drkmen) [inline spec/controllers/payments_controller_spec.rb]: it's still params
why called 'service'?

File: spec/support/factory_bot.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+RSpec.configure do |config|
+  config.include FactoryBot::Syntax::Methods
+end
```
Comment (drkmen) [inline spec/support/factory_bot.rb:5]: i guess this should be in spec_helper, is not it?

**Review by drkmen — CHANGES_REQUESTED** [review]

File: spec/support/factory_bot.rb (hunk 9 above)
Comment (magas4891) [reply spec/support/factory_bot.rb:5]: I can't remember why I did it, but based on the [article](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#rspec), this file should be, but you need to redo rails_helper a little.
i will do like this, ok?

File: spec/controllers/payments_controller_spec.rb (hunk 8 above)
Comment (magas4891) [reply spec/controllers/payments_controller_spec.rb]: hmm, why i changed it?
fixed

File: app/services/connections/family_connections/update_service.rb (hunk 5 above)
Comment (magas4891) [reply app/services/connections/family_connections/update_service.rb]: thanks, in the future I will not hesitate

File: app/controllers/users/invitations_controller.rb (hunk 4 above)
Comment (magas4891) [reply app/controllers/users/invitations_controller.rb]: done

File: app/controllers/family_connections_controller.rb (hunk 1 above)
Comment (magas4891) [reply app/controllers/family_connections_controller.rb]: +

File: app/controllers/users/invitations_controller.rb (hunk 3 above)
Comment (magas4891) [reply app/controllers/users/invitations_controller.rb]: it's clear

File: app/services/connections/family_connections/update_service.rb (hunk 6 above)
Comment (magas4891) [reply app/services/connections/family_connections/update_service.rb]: removed from here

File: app/controllers/family_connections_controller.rb (hunk 2 above)
Comment (magas4891) [reply app/controllers/family_connections_controller.rb]: permited

File: app/services/family_members/accept_invitation.rb (hunk 7 above)
Comment (magas4891) [reply app/services/family_members/accept_invitation.rb]: fixed

**Review by drkmen — APPROVED** [review]

**Pull request merged**